### PR TITLE
srravich/1973-Ngrok-Status-Viewer UX/UI fixes

### PR DESCRIPTION
**Pull request by srinaath**
This PR wraps up all the UX interactions for Ngrok Status Viewer.
- Shows flashing error indicator and opens the status viewer when there is a tunnel error
- Adds Status viewer to context menus for Mac and Windows
- Moves all the tunnel pings and timer updates into its own ngrokSagas

### CODE REVIEW

File: packages/app/main/src/ngrok.ts [hunk 1]
```diff
@@ -105,29 +107,34 @@ export class NgrokInstance {
     }
     await this.getNgrokInspectUrl(options);
     const tunnelInfo: { url; inspectUrl } = await this.runTunnel(options);
-    this.intervalForHealthCheck = setInterval(() => this.boundCheckTunnelStatus(tunnelInfo.url), 60000);
+    this.checkTunnelStatus();
+    this.intervalForHealthCheck = setInterval(() => this.boundCheckTunnelStatus(false), intervalForEachPing);
```
Comment (tonyanziano) [inline packages/app/main/src/ngrok.ts]: `checkTunnelStatus()` doesn't take any arguments in its signature, is the `false` arg needed?

File: packages/app/main/src/ngrokService.ts [hunk 2]
```diff
@@ -229,7 +229,10 @@ export class NgrokService {
     if (this.spawnErr) {
       emulatorApplication.mainWindow.logService.logToChat(
         conversationId,
-        textItem(LogLevel.Error, 'Failed to spawn ngrok'),
+        textItem(
+          LogLevel.Error,
+          'Failed to spawn ngrok. Please go to Ngrok Status Viewer and download the log file for more detailed view of the error.'
```
Comment (tonyanziano) [inline packages/app/main/src/ngrokService.ts]: `"Please go to the Ngrok Status Viewer and download the log file for a more detailed view of the error."`

File: packages/app/main/src/utils/fetchWithTimeout.spec.ts [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+        // Should not hit
+        expect(true).toBeFalsy();
+      })
+      .catch(er => {
+        expect(er).toBeDefined();
+        done();
+      });
+    jest.advanceTimersByTime(40001);
+  });
+});
```
Comment (tonyanziano) [inline packages/app/main/src/utils/fetchWithTimeout.spec.ts]: would it be easy to also test the fetch timing out and rejecting?

File: packages/app/main/src/state/sagas/ngrokSagas.ts [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+
+export const intervalForEachPing = 60000;
+export const intervalForRefreshedStatus = 20000;
+export const tunnelPingTimeout = 59000;
+
+const getPublicUrl = (state: RootState): string => state.ngrokTunnel.publicUrl;
+
+const getStaleNotificationIds = (state: RootState): string[] => state.ngrokTunnel.ngrokNotificationIds;
+
+// const getCurrentTunnelStatus = (state: RootState): TunnelStatus => state.ngrokTunnel.tunnelStatus;
```
Comment (tonyanziano) [inline packages/app/main/src/state/sagas/ngrokSagas.ts]: remove?

File: packages/app/main/src/state/sagas/ngrokSagas.ts [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+    if (isErrorResponse) {
+      return {
+        text: await response.text(),
+        status: response.status,
+      };
+    }
+    return undefined;
+  } catch (ex) {
+    return {
+      text: 'Tunnel ping has taken more than acceptable time limit. Looks like it does not exist anymore.',
```
Comment (tonyanziano) [inline packages/app/main/src/state/sagas/ngrokSagas.ts]: `"Tunnel ping has surpassed the acceptable time limit. Looks like it does not exist anymore."`

File: packages/app/main/src/state/sagas/ngrokSagas.ts [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+        if (ids.length > 0) {
+          yield all(ids.map((id: string) => call([commandService, commandService.remoteCall], Remove, id)));
+        }
+        action.payload.onTunnelPingSuccess();
+      }
+      yield put(setTimeIntervalSinceLastPing(TunnelCheckTimeInterval.Now));
+      yield delay(intervalForRefreshedStatus);
+      yield put(setTimeIntervalSinceLastPing(TunnelCheckTimeInterval.FirstInterval));
+      yield delay(intervalForRefreshedStatus);
+      // yield call(delay, intervalForRefreshedStatus);
```
Comment (tonyanziano) [inline packages/app/main/src/state/sagas/ngrokSagas.ts]: remove?

File: packages/app/client/src/ui/styles/themes/light.css [hunk 7]
```diff
@@ -291,4 +291,10 @@ html {
   --menu-item-hover-outline: none;
   --menu-separator-color: #2B2B2B;
   --menu-submenu-top-adjustment: -8px;
+
+  /* Ngrok Colors */
+  --ngrok-inactive: #007ACC;
+  --ngrok-active: #47B07F;
+  --ngrok-error: #BE1100;
+  --ngrok-error-outline: #F5B1B1;
```
[the remark this answers is not in the document]
Comment (srinaath) [reply packages/app/client/src/ui/styles/themes/light.css]: Added to high contrast css

File: packages/app/main/src/main.ts [hunk 8]
```diff
@@ -258,35 +260,25 @@ class EmulatorApplication {
...
-    }
-    const genericTunnelError =
-      'Oops.. Your ngrok tunnel seems to have an error. Please check the Ngrok Debug Console for more details';
+    const { Commands } = SharedConstants;
     dispatch(updateTunnelError({ ...response }));
 
     const ngrokNotification: Notification = newNotification(genericTunnelError);
+    dispatch(addNotification(ngrokNotification.id));
+
+    this.commandService.call(Commands.Ngrok.OpenStatusViewer, false);
```
[the remark this answers is not in the document]
Comment (srinaath) [reply packages/app/main/src/main.ts]: As discussed offline we dont need to await it

File: packages/app/main/src/main.ts [hunk 9]
```diff
@@ -258,35 +260,25 @@ class EmulatorApplication {
...
       this.commandService.remoteCall(Commands.Notifications.Remove, ngrokNotification.id);
-      dispatch(
-        EditorActions.open({
-          contentType: SharedConstants.ContentTypes.CONTENT_TYPE_NGROK_DEBUGGER,
-          documentId: SharedConstants.DocumentIds.DOCUMENT_ID_NGROK_DEBUGGER,
-          isGlobal: true,
-          meta: null,
-        })
-      );
+      this.commandService.call(Commands.Ngrok.OpenStatusViewer);
```
[the remark this answers is not in the document]
Comment (srinaath) [reply packages/app/main/src/main.ts]: Same as above

File: packages/app/main/src/utils/fetchWithTimeout.spec.ts (hunk 3 above)
Comment (tonyanziano) [reply packages/app/main/src/utils/fetchWithTimeout.spec.ts]: Ah, the test case name threw me off:

`should return successful response body` should maybe be something else like `should throw on timeout`

Comment (srinaath) [reply packages/app/main/src/utils/fetchWithTimeout.spec.ts]: Yep. i see y there was a confusion now. the test name def needs changing

**Pull request merged**